src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            await this.DownloadFileContainersAsync(downloadParameters.ProjectId, buildArtifacts, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public async Task DownloadFileContainersAsync(Guid projectId, IEnumerable<BuildArtifact> buildArtifacts, string targetDirectory, IEnumerable<string> minimatchFilters, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                await DownloadFileContainerAsync(projectId, buildArtifact, dirPath, minimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+        }
+
+        private (long, string) ParseContainerId(string resourceData)
+        {
+            // Example of resourceData: "#/7029766/artifacttool-alpine-x64-Debug"
+            var segments = resourceData.Split('/');
+            long containerId;
+            if (segments.Length < 3)
+            {
+                throw new ArgumentException($"Resource data value '{resourceData}' is invalid.");
+            }
+            if (segments.Length >= 3 && segments[0] == "#" && long.TryParse(segments[1], out containerId))
+            {
+                var artifactName = String.Join('/', segments[2]);
+                return(
+                        containerId,
+                        artifactName
+                    );
+            }
+            else
+            {
+                var message = $"Resource data value '{resourceData}' is invalid.";
+                throw new ArgumentException(message, "resourceData");
+            }
+        }
+
+        private async Task DownloadFileContainerAsync(Guid projectId, BuildArtifact artifact, string rootPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken, bool isSingleArtifactDownload=true)
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+            
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            tracer.Info($"Start downloading FCS artifact- {artifact.Name}");
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, tracer);
+            if (minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs, containerIdAndRoot.Item2);
+            }
+
+            if(!isSingleArtifactDownload && items.Any())
+            {
+                Directory.CreateDirectory(rootPath);
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
```

Comment:
Hi @omeshp , does FCS support downloading empty directories? If no, then I don't think we need this loop, because of lines 110-111.

Reply:
Yes, it does. We should create empty directories too since some customers depend on it.